DPDK cannot be installed on FreeBSD 14

FreeBSD 14 bus interface has been changed, which makes DPDK (through SPDK) unable to be installed. Gives the following error:

```
[...]kernel/freebsd/nic_uio/nic_uio.c:84:81: error: too many arguments provided to function-like macro invocation
DRIVER_MODULE(nic_uio, pci, nic_uio_driver, nic_uio_devclass, nic_uio_modevent, 0);

/usr/src/sys/sys/bus.h:832:9: note: macro 'DRIVER_MODULE' defined here
#define DRIVER_MODULE(name, busname, driver, evh, arg)
```

See also: https://forums.freebsd.org/threads/cannot-compile-old-trng-driver-under-freebsd-14.91171/

The patch below fixes this:

```
diff --git a/kernel/freebsd/nic_uio/nic_uio.c b/kernel/freebsd/nic_uio/nic_uio.c
index 7a81694c92..d052549701 100644
--- a/kernel/freebsd/nic_uio/nic_uio.c
+++ b/kernel/freebsd/nic_uio/nic_uio.c
@@ -78,10 +78,9 @@ struct pci_bdf {
        uint32_t function;
 };

-static devclass_t nic_uio_devclass;

 DEFINE_CLASS_0(nic_uio, nic_uio_driver, nic_uio_methods, sizeof(struct nic_uio_softc));
-DRIVER_MODULE(nic_uio, pci, nic_uio_driver, nic_uio_devclass, nic_uio_modevent, 0);
+DRIVER_MODULE(nic_uio, pci, nic_uio_driver, nic_uio_modevent, 0);

 static int
 nic_uio_mmap(struct cdev *cdev, vm_ooffset_t offset, vm_paddr_t *paddr,
```

--- a/subprojects/packagefiles/spdk/meson.build
+++ b/subprojects/packagefiles/spdk/meson.build
@@ -1,7 +1,7 @@
 project(
   'spdk',
   'c',
-  version: '24.01',
+  version: '24.05',
 )
 fs = import('fs')
 
@@ -112,7 +112,9 @@ spdk_libnames = get_option('build_subprojects') ? [
   'spdk_jsonrpc',
   'spdk_json',
   'spdk_util',
-  'spdk_log'
+  'spdk_log',
+  'spdk_dma',
+  'spdk_keyring'
 ] : []
 
 if get_option('build_subprojects') and host_machine.system() != 'freebsd'

--- a/subprojects/packagefiles/spdk/patches/spdk-0001-lib-bdev-Add-spdk_bdev_is_kv-to-bdev-interface.patch
+++ b/subprojects/packagefiles/spdk/patches/spdk-0001-lib-bdev-Add-spdk_bdev_is_kv-to-bdev-interface.patch
@@ -1,4 +1,4 @@
-From c63e4157923080d62b283bd51241174b5e943dc0 Mon Sep 17 00:00:00 2001
+From e70052dbb79411c36cc1be0ae3f71a09f7355915 Mon Sep 17 00:00:00 2001
 From: Karl Bonde Torp <k.torp@samsung.com>
 Date: Tue, 19 Sep 2023 10:08:29 +0200
 Subject: [PATCH 1/4] lib/bdev: Add spdk_bdev_is_kv() to bdev interface
@@ -18,10 +18,10 @@ Signed-off-by: Karl Bonde Torp <k.torp@samsung.com>
  4 files changed, 22 insertions(+)
 
 diff --git a/include/spdk/bdev.h b/include/spdk/bdev.h
-index 02fc06590..a1f78d355 100644
+index aa28b505d..6544171df 100644
 --- a/include/spdk/bdev.h
 +++ b/include/spdk/bdev.h
-@@ -687,6 +687,14 @@ bool spdk_bdev_is_md_interleaved(const struct spdk_bdev *bdev);
+@@ -746,6 +746,14 @@ bool spdk_bdev_is_md_interleaved(const struct spdk_bdev *bdev);
   */
  bool spdk_bdev_is_md_separate(const struct spdk_bdev *bdev);
  
@@ -37,10 +37,10 @@ index 02fc06590..a1f78d355 100644
   * Checks if bdev supports zoned namespace semantics.
   *
 diff --git a/include/spdk/bdev_module.h b/include/spdk/bdev_module.h
-index b5dd04bb4..92a1ae5f1 100644
+index 4ca3feeb9..8ffdd3b80 100644
 --- a/include/spdk/bdev_module.h
 +++ b/include/spdk/bdev_module.h
-@@ -558,6 +558,11 @@ struct spdk_bdev {
+@@ -563,6 +563,11 @@ struct spdk_bdev {
  	 */
  	uint32_t dif_check_flags;
  
@@ -53,10 +53,10 @@ index b5dd04bb4..92a1ae5f1 100644
  	 * Specify whether bdev is zoned device.
  	 */
 diff --git a/lib/bdev/bdev.c b/lib/bdev/bdev.c
-index 8daa452f0..e0008fd32 100644
+index 88d01d4ee..cecc7a74e 100644
 --- a/lib/bdev/bdev.c
 +++ b/lib/bdev/bdev.c
-@@ -4777,6 +4777,12 @@ spdk_bdev_is_md_separate(const struct spdk_bdev *bdev)
+@@ -4809,6 +4809,12 @@ spdk_bdev_is_md_separate(const struct spdk_bdev *bdev)
  	return (bdev->md_len != 0) && !bdev->md_interleave;
  }
  
@@ -70,10 +70,10 @@ index 8daa452f0..e0008fd32 100644
  spdk_bdev_is_zoned(const struct spdk_bdev *bdev)
  {
 diff --git a/lib/nvmf/ctrlr.c b/lib/nvmf/ctrlr.c
-index 5931303f0..eb585a00c 100644
+index e78cf1a30..c5adba1bc 100644
 --- a/lib/nvmf/ctrlr.c
 +++ b/lib/nvmf/ctrlr.c
-@@ -3116,6 +3116,9 @@ nvmf_ctrlr_identify_iocs(struct spdk_nvmf_ctrlr *ctrlr,
+@@ -3188,6 +3188,9 @@ nvmf_ctrlr_identify_iocs(struct spdk_nvmf_ctrlr *ctrlr,
  		if (spdk_bdev_is_zoned(ns->bdev)) {
  			vector->zns = 1;
  		}

--- a/subprojects/packagefiles/spdk/patches/spdk-0002-nvmf-Support-adding-KV-NS-to-nvmf_tgt.patch
+++ b/subprojects/packagefiles/spdk/patches/spdk-0002-nvmf-Support-adding-KV-NS-to-nvmf_tgt.patch
@@ -1,4 +1,4 @@
-From b098eb7ddc8803d506a8636a6fe968e55ad00807 Mon Sep 17 00:00:00 2001
+From 07e8c86d555654c05c7763991a6724cd154b7c77 Mon Sep 17 00:00:00 2001
 From: Karl Bonde Torp <k.torp@samsung.com>
 Date: Tue, 12 Sep 2023 08:16:44 +0200
 Subject: [PATCH 2/4] nvmf: Support adding KV NS to nvmf_tgt
@@ -29,10 +29,10 @@ index db9e5dc6d..ec21ced1e 100644
  		SPDK_WARNLOG("Unsupported CSI: %u for NSID: %u\n", ns->csi, ns->id);
  		return false;
 diff --git a/module/bdev/nvme/bdev_nvme.c b/module/bdev/nvme/bdev_nvme.c
-index 9cfd473e2..d84f1fe53 100644
+index 37b2bccf6..d6f49a9eb 100644
 --- a/module/bdev/nvme/bdev_nvme.c
 +++ b/module/bdev/nvme/bdev_nvme.c
-@@ -4013,6 +4013,10 @@ nvme_disk_create(struct spdk_bdev *disk, const char *base_name,
+@@ -4118,6 +4118,10 @@ nvme_disk_create(struct spdk_bdev *disk, const char *base_name,
  		disk->max_open_zones = spdk_nvme_zns_ns_get_max_open_zones(ns);
  		disk->max_active_zones = spdk_nvme_zns_ns_get_max_active_zones(ns);
  		break;

--- a/subprojects/packagefiles/spdk/patches/spdk-0003-lib-nvmf-Add-I-O-cmd-passthru-for-KV.patch
+++ b/subprojects/packagefiles/spdk/patches/spdk-0003-lib-nvmf-Add-I-O-cmd-passthru-for-KV.patch
@@ -1,4 +1,4 @@
-From f55f6d672b0ffff718c23b267b137c34d197c8d9 Mon Sep 17 00:00:00 2001
+From c7c8622c386791ea3115a7e9145cbb8e1f00b62e Mon Sep 17 00:00:00 2001
 From: Karl Bonde Torp <k.torp@samsung.com>
 Date: Fri, 15 Sep 2023 10:10:25 +0200
 Subject: [PATCH 3/4] lib/nvmf: Add I/O cmd passthru for KV
@@ -13,10 +13,10 @@ Signed-off-by: Karl Bonde Torp <k.torp@samsung.com>
  2 files changed, 8 insertions(+), 1 deletion(-)
 
 diff --git a/lib/nvmf/ctrlr.c b/lib/nvmf/ctrlr.c
-index eb585a00c..1288e77bd 100644
+index c5adba1bc..c0c137f84 100644
 --- a/lib/nvmf/ctrlr.c
 +++ b/lib/nvmf/ctrlr.c
-@@ -345,7 +345,7 @@ nvmf_subsys_has_multi_iocs(struct spdk_nvmf_subsystem *subsystem)
+@@ -401,7 +401,7 @@ nvmf_subsys_has_multi_iocs(struct spdk_nvmf_subsystem *subsystem)
  
  	for (i = 0; i < subsystem->max_nsid; i++) {
  		ns = subsystem->ns[i];
@@ -25,7 +25,7 @@ index eb585a00c..1288e77bd 100644
  			return true;
  		}
  	}
-@@ -4336,6 +4336,8 @@ nvmf_ctrlr_process_io_cmd(struct spdk_nvmf_request *req)
+@@ -4400,6 +4400,8 @@ nvmf_ctrlr_process_io_cmd(struct spdk_nvmf_request *req)
  	if (spdk_nvmf_request_using_zcopy(req)) {
  		assert(req->zcopy_phase == NVMF_ZCOPY_PHASE_INIT);
  		return nvmf_bdev_ctrlr_zcopy_start(bdev, desc, ch, req);
@@ -35,11 +35,11 @@ index eb585a00c..1288e77bd 100644
  		switch (cmd->opc) {
  		case SPDK_NVME_OPC_READ:
 diff --git a/lib/nvmf/subsystem.c b/lib/nvmf/subsystem.c
-index b93257262..31ebfd5d0 100644
+index 07340b89c..fd1280b96 100644
 --- a/lib/nvmf/subsystem.c
 +++ b/lib/nvmf/subsystem.c
-@@ -1841,6 +1841,11 @@ spdk_nvmf_subsystem_add_ns_ext(struct spdk_nvmf_subsystem *subsystem, const char
- 		subsystem->max_zone_append_size_kib = max_zone_append_size_kib;
+@@ -2141,6 +2141,11 @@ spdk_nvmf_subsystem_add_ns_ext(struct spdk_nvmf_subsystem *subsystem, const char
+ 		}
  	}
  
 +	if (spdk_bdev_is_kv(ns->bdev)) {

--- a/subprojects/packagefiles/spdk/patches/spdk-0004-lib-nvmf-Add-identify_-_iocs-passthru-for-KV.patch
+++ b/subprojects/packagefiles/spdk/patches/spdk-0004-lib-nvmf-Add-identify_-_iocs-passthru-for-KV.patch
@@ -1,4 +1,4 @@
-From 98949e7bbe035bd829c2cd3a0a51c26ed0f56dba Mon Sep 17 00:00:00 2001
+From 2d74adffa39c3aa342756474fc716c07328d264e Mon Sep 17 00:00:00 2001
 From: Karl Bonde Torp <k.torp@samsung.com>
 Date: Mon, 25 Sep 2023 15:59:24 +0200
 Subject: [PATCH 4/4] lib/nvmf: Add identify_*_iocs passthru for KV
@@ -9,7 +9,7 @@ Signed-off-by: Karl Bonde Torp <k.torp@samsung.com>
  1 file changed, 11 insertions(+), 2 deletions(-)
 
 diff --git a/lib/nvmf/ctrlr.c b/lib/nvmf/ctrlr.c
-index 1288e77bd..97020575b 100644
+index c0c137f84..0762ce1d8 100644
 --- a/lib/nvmf/ctrlr.c
 +++ b/lib/nvmf/ctrlr.c
 @@ -55,6 +55,7 @@ static struct spdk_nvmf_custom_admin_cmd g_nvmf_custom_admin_cmd_hdlrs[SPDK_NVME
@@ -20,7 +20,7 @@ index 1288e77bd..97020575b 100644
  
  static inline void
  nvmf_invalid_connect_response(struct spdk_nvmf_fabric_connect_rsp *rsp,
-@@ -3176,10 +3177,18 @@ nvmf_ctrlr_identify(struct spdk_nvmf_request *req)
+@@ -3248,10 +3249,18 @@ nvmf_ctrlr_identify(struct spdk_nvmf_request *req)
  				tmpbuf, req->length);
  		break;
  	case SPDK_NVME_IDENTIFY_NS_IOCS:

--- a/subprojects/spdk.wrap
+++ b/subprojects/spdk.wrap
@@ -4,7 +4,7 @@
 
 [wrap-git]
 url = https://github.com/spdk/spdk.git
-revision = v24.01
+revision = v24.05
 patch_directory = spdk
 clone-recursive = true
 

